spec/Appendix A -- Persisted Documents.md

```diff
@@ -95,6 +95,35 @@ implementation specific.
 Note: A 32 character hexadecimal _custom document identifier_ is likely to be an
 MD5 hash of the GraphQL document, as traditionally used by Relay.
 
+### Document identifier syntax
+
+DocumentIdentifier ::
+
+- PrefixedDocumentIdentifier
+- CustomDocumentIdentifier
+
+PrefixedDocumentIdentifier ::
+
+- Sha256HexDocumentIdentifier
+- OtherPrefixedDocumentIdentifier
+
+Sha256HexDocumentIdentifier : `sha256:` Sha256Checksum
+
+Sha256Checksum :: 64 LowerCaseHexDigit
+
+LowerCaseHexDigit :: one of
+
+- `0` `1` `2` `3` `4` `5` `6` `7` `8` `9`
+- `a` `b` `c` `d` `e` `f`
+
+OtherPrefixedDocumentIdentifier :: `x-` IdentifierCharacter+ `:` IdentifierCharacter+
+
+CustomDocumentIdentifier :: SourceCharacter+
+
+IdentifierCharacter :: SourceCharacter but not `:`
+
+SourceCharacter :: Any Unicode scalar value
```

Comment:
This is incredibly broad; I know that's effectively what we have currently but I feel like we should lock it down a bit.

Reply:
While we could allow all Unicode characters, the main spec defines a specific set of characters for use in that document.  Something similar could be done here. 

[SourceCharacter](https://spec.graphql.org/October2021/#SourceCharacter)
U+0009
U+000A
U+000D
U+0020–U+FFFF

On a side note, the spec is defining these characters as UTF-16 codes, not Unicode code points. 

I’m fine with allowing any Unicode character in this context, or any non control character would be fine too.

Reply:
I had only alphanumericals initially and then backtracked to "any unicode" because "why not". But I think your point about having the identifier in url path is the important one. I'll move to alphanumerical + a bunch of unreserved chars like you did [here](https://github.com/graphql/graphql-over-http/commit/a3d9ecb9e988a22cb9010618becbc67117710ec8). Questions: 

* do we even need `~`, `.`, `_`, `-`?
* do we need uppercase letters? This might cause issues on case-insensitive filesystems (hello macOS 👋 )

Reply:
I strongly recommend a wide range of characters. URLs can escape any Unicode character so there need be no limit on what characters can be used for document ids. Also, url paths are by definition case sensitive; but they may be treated with no case sensitivity. For instance, if the document id was a guid it would need dashes. If it was base64 it would need lowercase and uppercase and a couple extra characters. If it was readable names, it would need to support Unicode for foreign languages. And a custom implementation may decide to treat the document id with no case sensitivity.

Reply:
> if the document id was a guid it would need dashes

Excellent point.

> If it was base64 it would need lowercase and uppercase and a couple extra characters

Another excellent point. `+`, `/` and `=` would be needed too. Which would make them a bit awkard to use as url path segments... Interesting!

> If it was readable names, it would need to support Unicode for foreign languages

I think I disagree on this one. I don't  really see anyone using readable names as identifiers? Sounds like a footgun to me that if we can, we should discourage. Just like the relay cursors are encouraged to be opaque, I would encourage the document ids to be similarly opaque.

Tangential: I have a GraphQL-over-HTTP meeting in my calendar on Thursday. Is that still happening? Want to discuss over there?

Reply:
Personally, the most I would exclude is control characters, colon, and the two slashes. Or otherwise similar to most OS filename conventions.

Reply:
My current mental model is:

* OS filenames are user interface. They are visible to the user and should be human readable. 
* POs identifiers are technical and should be machine readable. 

I don't really see myself using french diacritics in my POs ids (although that could make for a fun API!). But again all the POs I have ever used are sha256 so maybe there are other use cases out there.

Reply:
> Another excellent point. +, / and = would be needed too. Which would make them a bit awkard to use as url path segments... Interesting!

There's standard URL-safe base64 encoding that uses `-` and `_`: https://datatracker.ietf.org/doc/html/rfc4648#section-5 (the padding can be omitted).

> I think I disagree on this one. I don't really see anyone using readable names as identifiers? Sounds like a footgun to me that if we can, we should discourage. 

Very strong agree that persisted document IDs should not be hand-crafted. They should be hashes or counters or similar - computer assigned.

>  I have a GraphQL-over-HTTP meeting in my calendar on Thursday. Is that still happening? Want to discuss over there?

I've just created agenda files for them so feel free to add yourself if you want :+1: 

> Personally, the most I would exclude is control characters, colon, and the two slashes. Or otherwise similar to most OS filename conventions.

I don't see much value in this. In general it's best to be strict first and loosen up when a need arises, since you cannot become stricter later without it being a breaking change. What use case do you have for being defined loosely? Our intended use cases are GUIDs/UUIDs, hashes and counters, and those are typically numeric, hexadecimal, or base64 encoded. I can imagine namespacing them too (e.g. `MyApp.f9db3fbb3cfe...`) but other than the a  I don't think they should be much wider than a GraphQL `Name`.

Reply:
👍 I added my name [for Thursday's meeting](https://github.com/graphql/graphql-over-http/pull/298)

Reply:
> What use case do you have for being defined loosely?

Personally, none at all. Just nowhere else are there similar restrictions on identifiers. Filenames have a very few restrictions. And here we can’t use colon or the string would be confused with a prefixed identifier. So I would wonder why we are further restricting these identifiers which are otherwise opaque to the application?

Comparing to GraphQL names, there it’s part of a formatted document where a parser must be able to read the identifier separate from other tokens. But here all these identifiers are already escaped by JSON just like any other string. As part of a url path, again there is a well defined encoding scheme available. 

Personally, I’d probably use url-safe base64 and period or fewer characters.

Reply:
> Very strong agree that persisted document IDs should not be hand-crafted. They should be hashes or counters or similar - computer assigned.

I just want to point out that part of the discussion during the 7-25-2024 working group was whether we should add a recommendation to include the document identifier and operation name within the url, specifically to make it easier to be human-readable in logs and during debugging.  Given that many programmers do not utilize operation names, I think it is reasonable to guess that a valid document identifier naming scheme may be a combination of a description and hash/id, similar to machine-generated human-readable SEO product identifiers.

For instance, the algorithm may take the filename of the tsx file containing the query along with 6 base64 characters of the hash (48 bits-worth) for something like `edit_product~3K-A2J-` as the document identifier.

Reply:
I attempted an implementation of persisted documents today.  The client is using [encodeUriComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) to encode the document Id into the url in a format such as `/graphql?documentId=sha256:abcdef...`.  I immediately found that `:` was encoded as `%3A` due to the potentical conflict with `://`.  In short, `:` is not a url-safe character.  As such, it's somewhat pointless to argue that we should only allow URL-safe characters in the payload when the document id cannot possibly be URL-safe as we have defined `:`, a most-assuredly non-url-safe character, is the separator between a document id's prefix and its payload.

If we wish for the document ids to be url-safe, we should pick another character (such as `.`) as the separator between prefix and payload.  Using `.` would still allow Base64Url (which uses `_` and `-`).  Or we could use another character such as `!` so that `.` is available for custom identifiers that wish to use `.`.

Reply:
Good catch @Shane32; `:` sits in that grey area where you don't strictly need to escape it, but it's best practice to do so. We should change that to a URL-safe character.

Reply:
I like the `!` suggestion best I think. I think we can change the spec text to that without making existing `:` usage non-compliant - it just becomes a custom ID rather than prefixed ID.

Reply:
`*` seems to work with both `URLSearchParams` and `encodeURIComponent`, and I don't think it's normally used in encodings so should be safe... I'll need to look up in the relevant specs to see if it's not meant to be used for some reason.

Reply:
Spec for URLSearchParams:

> The [application/x-www-form-urlencoded percent-encode set](https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set) contains all code points, except the [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), U+002A (*), U+002D (-), U+002E (.), and U+005F (_).

So `~` is probably encoded also.